File: src/main/java/itmo/is/project/service/module/production/ProductionModuleBuildService.java
package itmo.is.project.service.module.production;

import itmo.is.project.dto.ProductionModuleBlueprintDto;
import itmo.is.project.dto.ProductionModuleDto;
import itmo.is.project.dto.request.ModuleBuildRequest;
import itmo.is.project.mapper.module.ProductionModuleBlueprintMapper;
import itmo.is.project.mapper.module.ProductionModuleMapper;
import itmo.is.project.model.module.production.ProductionModule;
import itmo.is.project.model.module.production.ProductionModuleBlueprint;
import itmo.is.project.repository.module.production.ProductionModuleBlueprintRepository;
import itmo.is.project.repository.module.production.ProductionModuleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ProductionModuleBuildService {
    private final ProductionModuleRepository productionModuleRepository;
    private final ProductionModuleMapper productionModuleMapper;

    private final ProductionModuleBlueprintRepository productionModuleBlueprintRepository;
    private final ProductionModuleBlueprintMapper productionModuleBlueprintMapper;

    public Page<ProductionModuleBlueprintDto> findAllBlueprints(Pageable pageable) {
        return productionModuleBlueprintRepository.findAll(pageable)
                .map(productionModuleBlueprintMapper::toDto);
    }

    public ProductionModuleDto buildModule(ModuleBuildRequest moduleBuildRequest) {
        Integer blueprintId = moduleBuildRequest.blueprintId();
        ProductionModuleBlueprint blueprint = productionModuleBlueprintRepository.findById(blueprintId).orElseThrow();
        ProductionModule productionModule = new ProductionModule();
        productionModule.setBlueprint(blueprint);
        productionModule = productionModuleRepository.save(productionModule);
        return productionModuleMapper.toDto(productionModule);
    }
}

File: src/main/kotlin/itmo/isproject/production/service/ProductionModuleBuildService.kt
package itmo.isproject.production.service

import itmo.isproject.shared.module.service.BuildService
import itmo.isproject.production.dto.ProductionModuleBlueprintDto
import itmo.isproject.production.dto.ProductionModuleDto
import itmo.isproject.production.mapper.ProductionModuleBlueprintMapper
import itmo.isproject.production.mapper.ProductionModuleMapper
import itmo.isproject.production.model.ProductionModule
import itmo.isproject.production.model.ProductionModuleBlueprint
import itmo.isproject.production.repository.ProductionModuleBlueprintRepository
import itmo.isproject.production.repository.ProductionModuleRepository
import itmo.isproject.storage.api.StorageService
import org.springframework.stereotype.Service

@Service
class ProductionModuleBuildService(
    storageService: StorageService,
    productionModuleBlueprintRepository: ProductionModuleBlueprintRepository,
    productionModuleBlueprintMapper: ProductionModuleBlueprintMapper,
    productionModuleRepository: ProductionModuleRepository,
    productionModuleMapper: ProductionModuleMapper
) : BuildService<ProductionModule, ProductionModuleBlueprint, ProductionModuleDto, ProductionModuleBlueprintDto>(
    storageService,
    productionModuleBlueprintRepository,
    productionModuleBlueprintMapper,
    productionModuleRepository,
    productionModuleMapper,
    ::ProductionModule
)